server/src/main/java/io/crate/execution/dml/delete/TransportShardDeleteAction.java

```diff
@@ -103,21 +106,23 @@ protected WritePrimaryResult<ShardDeleteRequest, ShardResponse> processRequestIt
                             logger.debug("shardId={} failed to execute delete for id={}, doc not found",
                                 request.shardId(), item.id());
                         }
-                        shardResponse.add(location,
-                            new ShardResponse.Failure(
-                                item.id(),
-                                "Document not found while deleting",
-                                false));
+                        errorCount++;
+                        Throwable ex = null;
+                        if (errorCount <= BULK_RESPONSE_MAX_ERRORS_PER_SHARD) {
+                            ex = new DocumentMissingException(indexShard.shardId(), Constants.DEFAULT_MAPPING_TYPE, item.id());
+                        }
+                        shardResponse.add(location, new ShardResponse.Failure(item.id(), ex, false));
```

Comment:
If I'm reading this right maybe the `errorCount` & `<=   BULK_RESPONSE_MAX_ERRORS_PER_SHARD`  handling could be moved into `shardResponse` to avoid some duplication between shardDeleteAction and shardUpsertAction and encapsulate it.

Not sure if worth it, but then we could even follow up in another PR to do some sort of de-duduplication - instead of adding new failure instances we keep a list of failed item id's with the same error or something like that.

Reply:
Good point, done via https://github.com/crate/crate/pull/16799/commits/1a186eb52a8c90b27155ce0ff443e490958d41d7